### PR TITLE
Use regex expression for authorization

### PR DESCRIPTION
**Pull request by Chaho12**
Use regex instead of contains to support more cases where more than 1 type of user members can exist and not all but some of the groups/members are allowed for User role.

### CODE REVIEW

File: docs/security.md [hunk 1]
```diff
@@ -72,39 +72,42 @@ authentication:
...
   admin: 'lb_admin'
   user: 'lb_user'
-  api: "lb_api"
-  ldapHost:
-  ldapPort:
-  ldapBindDn:
-  ldapSearch:
-  ldapPassword:
+  api: 'lb_api'
+  ldapConfigPath: <ldap_config_path>
```
Comment (willmostly) [inline docs/security.md:110]: This change doesn't appear to be related to the change from contains => matches. Were they incorrect?

File: docs/security.md [hunk 2]
```diff
@@ -72,39 +72,42 @@ authentication:
...
-users. User attributes from LDAP is supported or you can use the preset users
-defined in the yaml file. Authorization is supported via LDAP user attributes
+users. You can use the preset users defined in the yaml file. 
+LDAP Authorization is also supported by adding user attribute configs in file.
+
+- Check out [LDAPTestConfig.yml](https://github.com/trinodb/trino-gateway/blob/main/gateway-ha/src/test/resources/auth/ldapTestConfig.yml) file for config details
 
 ```
 authorization:
   admin: 'lb_admin'
```
Comment (willmostly) [inline docs/security.md:107]: A note should be added that this supports a regex

**Review by willmostly — COMMENTED** [review]
Logic looks good, just a couple notes on the docs and I think this will be ready.

File: docs/security.md (hunk 2 above)
Comment (Chaho12) [reply docs/security.md:107]: In line 90 I said `The roles in regex string format are supported by trino load balancer`, is this not enough?
Maybe like this?
```suggestion
# Value in regex format
authorization:
  admin: 'lb_admin'
```

File: docs/security.md (hunk 1 above)
Comment (Chaho12) [reply docs/security.md:110]: Ah yeah, it's a config path (string) that is read as ldap config has serveral properties required fields.

Comment (Chaho12) [reply docs/security.md:110]: Since https://github.com/trinodb/trino-gateway/pull/36 PR has this part covered, i'll remove this change

File: docs/security.md (hunk 2 above)
Comment (willmostly) [reply docs/security.md:107]: Sorry, I missed that line when I added this comment. It probably doesn't hurt to repeat it here though.

File: gateway-ha/src/main/java/io/trino/gateway/ha/security/LbAuthorizer.java [hunk 3]
```diff
@@ -19,17 +19,17 @@ public boolean authorize(LbPrincipal principal, String role) {
       case "ADMIN":
         log.info("User {} identified as ADMIN", principal.getName());
         return principal.getMemberOf()
-            .filter(m -> m.contains(configuration.getAdmin()))
+            .filter(m -> m.matches(configuration.getAdmin()))
```
Comment (willmostly) [inline gateway-ha/src/main/java/io/trino/gateway/ha/security/LbAuthorizer.java:22]: There should be a test case for this functionality

File: docs/security.md (hunk 2 above)
Comment (Chaho12) [reply docs/security.md:108]: Added :)

File: gateway-ha/src/main/java/io/trino/gateway/ha/security/LbAuthorizer.java (hunk 3 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/security/LbAuthorizer.java:22]: I added test code. PTAL

File: docs/security.md [hunk 4]
```diff
@@ -72,39 +72,43 @@ authentication:
...
 ### Form/LDAP
 
+LDAP requires both random key pair and config path for LDAP
+
 ```
 authentication:
   defaultType: "form"
   form:
     ldapConfigPath: <ldap_config_path>
+    selfSignKeyPair:
```
Comment (vishalya) [inline docs/security.md:82]: nice!

**Review by vishalya — APPROVED** [review]
LGTM

**Comment by Chaho12** [comment]
Rebased and pushed to be ready for merge

**Review by mosabua — APPROVED** [review]
Looks good to me as well. 

Any further input @willmostly ?

**Comment by willmostly** [comment]
Sorry for the delay @Chaho12 - I had a review pending that I didn't realize never got submitted

**Comment by mosabua** [comment]
Could you rebase and adjust to review comments please @Chaho12 ?

File: gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    private static final String UNKNOWN_ROLE = "UNKNOWN";
+
+    private static final String PVFX_DATA_31 = "PVFX_DATA_31";
+
+    static void match(String member_of, String conf, String role, boolean expectedToMatch) {
+        AuthorizationConfiguration configuration = new AuthorizationConfiguration();
+        principal = new LbPrincipal(USER, Optional.of(member_of));
+        if (role.equalsIgnoreCase(ADMIN_ROLE)){
+            configuration.setAdmin(conf);
+            authorizer = new LbAuthorizer(configuration);
```
[the remark this answers is not in the document]
Comment (Chaho12) [reply gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java:29]: The `conf` that you mentioned is a regex string and used to create LbAuthorizer, which changes for every test cases. 
I don't quite understand when you mean `assign users to roles in a @BeforeClass`. 
User or `String member_of`(`PVFX_DATA_31`) is fixed for all tests so I could get rid of that in all functions.
What i intended was
- `assertMatch(PVFX_DATA_31, "PVFX(.*)", ADMIN_ROLE);` means that user with name PVFX_DATA_31 and gatewate conf  `PVFX(.*)` will be used to check with ADMIN_ROLE.

File: gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    @Test
+    public void testBasic() throws Exception {
+        assertMatch(PVFX_DATA_31, PVFX_DATA_31, ADMIN_ROLE);
+        assertNotMatch(PVFX_DATA_31, "PVFX", USER_ROLE);
+        assertNotMatch(PVFX_DATA_31, "DATA", API_ROLE);
+        assertNotMatch(PVFX_DATA_31, "31", UNKNOWN_ROLE);
+    }
+
+    @Test
+    public void testZeroOrMoreCharacters() throws Exception {
```
[the remark this answers is not in the document]
Comment (Chaho12) [reply gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java:81]: Is this ok?
https://github.com/trinodb/trino-gateway/pull/32/files#diff-edc26af1293dd63c7aaad0eeb78d264cd58afcb12df845fb8e3921aa24d89e02R82-R83

File: gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Chaho12) [reply gateway-ha/src/test/java/io/trino/gateway/ha/security/TestLbAuthorizer.java]: Oh now I understood what you meant when separting `conf` with `match`.
I changed as you told! 
I also logged more info when authorizing

**Review by willmostly — APPROVED** [review]
Looks good, thanks!
